lib/src/validator/pubspec_typo.dart

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import '../dice_coefficient.dart';
+import '../entrypoint.dart';
+import '../validator.dart';
+
+/// Validates that a package's pubspec does not contain any typos in its keys.
+class PubspecTypoValidator extends Validator {
+  PubspecTypoValidator(Entrypoint entrypoint) : super(entrypoint);
+
+  @override
+  Future validate() async {
+    final fields = entrypoint.root.pubspec.fields;
+
+    if (fields == null) return;
+
+    for (final key in fields.keys) {
+      var bestDiceCoefficient = 0.0;
+      var closestKey = '';
+
+      for (final validKey in _validPubspecKeys) {
+        final dice = diceCoefficient(key, validKey);
+        if (dice > bestDiceCoefficient) {
+          bestDiceCoefficient = dice;
+          closestKey = validKey;
+        }
+      }
+
+      // 0.73 is a magic value determined by looking at the most common typos
+      // in all the pubspecs on pub.dev.
+      if (bestDiceCoefficient >= 0.73 && bestDiceCoefficient < 1.0) {
+        warnings.add('$key is not a key recognizable by pub - '
+            'did you mean $closestKey?');
+      }
+
+      if (closestKey == 'author' || closestKey == 'authors') {
+        warnings.add('The "$closestKey" key is deprecated - Use a verified '
+            'publisher (https://dart.dev/tools/pub/verified-publishers) '
+            'instead.');
+      }
+    }
+  }
+}
+
+/// List of keys in `pubspec.yaml` that will be recognized by pub, as well as
+/// the depreciated `author` key.
+///
+/// Retrieved from https://dart.dev/tools/pub/pubspec
```

Comment:
Maybe we should write a tool in `tool/pubspec-field-histogram.dart` that iterates through all pubspecs on pub.dev and makes a histogram counting how frequently top-level keys are used.

We shouldn't print typo warnings for deprecated keys, and we kind of want to discover what people are using.

Reply:
Already done :+1: 

name: 14677
description: 14677
version: 14677
homepage: 14575
environment: 13656
dependencies: 12982
dev_dependencies: 11980
flutter: 7328
authors: 851
author: 8455
repository: 1190
issue_tracker: 521
documentation: 520
executables: 367
dependency_overrides: 77
scripts: 8
ffigen: 1
verified publisher: 4
targets: 1
publish_to: 58
analyzer: 12
license: 23
linter: 4
maintainer: 82
copyright: 3
email: 27
dartdoc: 5
custom: 3
dev_dependecies: 1
publisher: 4
publish: 1
respsitory: 2
auth: 1
_dependency_overrides: 3
import_sorter: 1
homeage: 1
Repository: 1
sdk: 3
doucumentation: 1
module: 8
plugin: 2
git: 13
deprecated: 2
deprecation_message: 2
api: 1
source: 1
Site: 1
docs: 1
dev_dependencies_overrides: 1
transformers: 256
changelog -x: 1
issues: 2
protocol: 128
imc_def: 1
repostory: 2
environment_config: 1
test: 1
strong-mode: 1
latest_commit: 1
assets: 2
enable-experiment: 2
docstring: 3
dartrix: 3
contributors: 1
reposity: 1
streams_channel: 1
uploader: 3
COPYING: 1
LICENSE: 1
UNLICENSE: 1
README.MD: 1
github: 9
Reviewed-by: 1
dependecies: 2
~ pedantic: 1
publishers: 2
flutter_icons: 2
flutter_manifest: 1
auther: 1
autho: 1
pubpage: 8
originauthor: 1
originhomepage: 1
env: 1
web: 9
uses-material-design: 1
flutter_i18n: 2
about: 1
website: 3
private: 1
path: 7
anukeerthi_library: 2
http: 1
excutables: 1
dev_dependency_overrides: 4
repositiory: 1
dev-dependencies: 2
publish-to: 1
environtment: 1
executable: 4
changelogname: 2
ghooks: 1
refit_http: 1
jaguar_http: 2
dev_transformers: 1
enviroent: 1
environnment: 4
devDependencies: 2
dev_dependency: 1

Reply:
Checking for the `author` key was suggested by @sigurdm, since it was a key that was previously recognized, and we are able to detect it and suggest the necessary corrective actions to be taken instead! I could remove it if you want though!